### ISSUE TITLE
doh: zero length hostname causes buffer read underflow

### ISSUE THREAD

**Issue by pauldreik** [curl/curl#4605]
https://github.com/curl/curl/commit/b6a53fff6c1d07e8a9cb65ca1066d99490fb8132 introduced a change which [reads the buffer](https://github.com/curl/curl/blob/b6a53fff6c1d07e8a9cb65ca1066d99490fb8132/lib/doh.c#L114) before [checking the index](https://github.com/curl/curl/blob/b6a53fff6c1d07e8a9cb65ca1066d99490fb8132/lib/doh.c#L120).

To reproduce: invoke doh_encode with an empty string.

This would have been caught by [the internal fuzzers](https://github.com/curl/curl-fuzzer/issues/32) (that is how I detected it now).

Perhaps it would be good to add an empty string to the unit test?

**Comment by bagder** [comment]
Hence the assert there in the code. But the question is how do you get the function called with a zero length name? I'm not aware of any code path leading there that can be zero length.

**Comment by pauldreik** [comment]
Even if it can not be called with zero length today, I think it is wise to protect against it. 
 - future refactoring may one day make that happen
 - defence in depth - protect against bugs or mistakes elsewhere
 - the fuzzer does not need special casing, improving the likelihood that it actually finds (future) bugs.
 - the fix is cheap

What do you think of the following patch?
```diff
diff --git a/lib/doh.c b/lib/doh.c
index 73ccce95e..ca77bc076 100644
--- a/lib/doh.c
+++ b/lib/doh.c
@@ -108,8 +108,10 @@ UNITTEST DOHcode doh_encode(const char *host,
    * the overall length by one.
    */
 
+  if(hostlen == 0)
+    return DOH_DNS_BAD_LABEL;
+
   size_t expected_len;
-  DEBUGASSERT(hostlen);
   expected_len = 12 + 1 + hostlen + 4;
   if(host[hostlen-1]!='.')
     expected_len++;
```

Ping @niallor

**Comment by bagder** [comment]
The assert is there to trigger if anyone would ever call the the function wrongly, as could be the result of a refactor. It would still be weird to call this function with a zero length host name - if there's ever a risk we should catch that problem earlier.

Having run-time code that protects against NNN implies that the NNN situation can indeed legitimately happen (and static code analyzers as well as human beings will consider it so). Having run-time code that logically can never be reached typically qualifies as "dead code".

I understand that just checking for a zero length and bail out on that is easy and quick, but I think it sets a bad precedence: adding run-time checks for situations that only ever occur when someone rips the curl code apart and calls it in ways we haven't designed the functions to be called.

Isn't it just about as easy to just make sure the fuzzer has at least one byte in the buffer when it calls the encode function? Surely it can't be the only function that has this requirement?

Finally: I thought the fuzzer was built with debug enabled so that the assert would trigger...

**Comment by pauldreik** [comment]
I agree with your arguments, thanks for motivating instead of just saying no. I guess we just value things differently!

> static code analyzers

I would expect (a very smart) static code analyzer to understand strlen() and the buf[len-1], but not the nonstandard DEBUGASSERT macro. I do not know if cppcheck and/or covertiy reports on this?

> Isn't it just about as easy to just make sure the fuzzer has at least one byte in the buffer when it calls the encode function? Surely it can't be the only function that has this requirement?

I think this is the first one I encounter in curl. Other situations may demand null terminated strings, which I have a [convenience function for](https://github.com/pauldreik/curl-fuzzer/blob/b5ec16f270f5737c4e06427c4eb4b4c9070267ae/intree_fuzzer/src/insidefuzzers/FuzzData.h#L52).

> Finally: I thought the fuzzer was built with debug enabled so that the assert would trigger...

It is, and the assert triggers.

**Comment by pauldreik** [comment]
[The fuzzer is updated](https://github.com/pauldreik/curl-fuzzer/commit/9b234e68c355c8509c75199a4a15c07dea2934cc), ran it for a few minutes without any problems.

**Comment by niallor** [comment]
On 17 Nov 2019, at 5:03, Paul Dreik wrote:

> What do you think of the following patch?
>
>     diff --git a/lib/doh.c b/lib/doh.c
>     index 73ccce95e..ca77bc076 100644
>     --- a/lib/doh.c
>     +++ b/lib/doh.c
>     @@ -108,8 +108,10 @@ UNITTEST DOHcode doh_encode(const char *host,
>         * the overall length by one.
>         */
>
>     +  if(hostlen == 0)
>     +    return DOH_DNS_BAD_LABEL;
>     +
>        size_t expected_len;
>     -  DEBUGASSERT(hostlen);
>        expected_len = 12 + 1 + hostlen + 4;
>        if(host[hostlen-1]!='.')
>          expected_len++;
>
> Ping [@niallor](<https://github.com/niallor>)

TL;DR: neither method seems to me either harmful or useful. 8-)

Since DOH is unaware of locally configured search suffixes which
might be used to give context to a relative domain name, an empty
hostname is formally equivalent to the hostname ".", which
isn't forbidden.  Either of these is legitimate as a DNS name, and
refers to the root node of the DNS namespace. This is quite validly
QNAME-encoded as a single octet with value '\0', so that the danger
of encoding a malformed QNAME doesn't arise, and need not be
defended against.

Defending against an invalid reference to `host[hostlen-1]`
might more simply be done by using
`((hostlen) && (host[hostlen-1] != '.'))`
as the if-condition. The consequent effect of not incrementing
`expected_len` would be correct, as there is no need to make the
adjustment in the case of an absent label.

Prevention of a use-case which libcurl refuses (or has no need) to
support makes some sense, but is incomplete without also forbidding
"." as the hostname.

Right now, I'm not aware of a compelling need for libcurl either
to support queries against the root node of the DNS, or to forbid
them.

I hope this helps.
/Niall
